src/zenml/orchestrators/step_run_utils.py

```diff
@@ -0,0 +1,466 @@
+        pipeline_run=pipeline_run
+    )
+
+    while (
+        cache_candidates := find_cacheable_invocation_candidates(
+            deployment=deployment,
+            finished_invocations=cached_invocations,
+        )
+        - visited_invocations
+    ):
+        for invocation_id in cache_candidates:
+            visited_invocations.add(invocation_id)
+
+            try:
+                step_run_request = request_factory.create_request(
+                    invocation_id
+                )
+                request_factory.populate_request(step_run_request)
+            except Exception as e:
+                # We failed to create/populate the step run. This might be due
+                # to some input resolution error, or an error importing the step
+                # source (there might be some missing dependencies). We just
+                # treat the step as not cached and let the orchestrator try
+                # again in a potentially different environment.
+                logger.debug(
+                    "Failed preparing step run `%s`: %s", invocation_id, str(e)
+                )
+                continue
+
+            if step_run_request.status != ExecutionStatus.CACHED:
+                # If we're not able to cache the step run, the orchestrator
+                # will run the step later which will create the step run
+                # -> We don't need to do anything here
+                continue
+
+            step_run = Client().zen_store.create_run_step(step_run_request)
+
+            # Refresh the pipeline run here to make sure we have the latest
+            # state
+            pipeline_run = Client().get_pipeline_run(pipeline_run.id)
+
+            step_model, step_run = prepare_step_run_model(
+                step_run=step_run, pipeline_run=pipeline_run
+            )
+
+            model = step_model or pipeline_model
+
+            # TODO: This function also imports the step code, which means it
+            # will fail when running a template. The only reason it imports
+            # the step is to load the artifact configs to check whether a
+            # different model is specified for a single output. We can go two
+            # ways here I think:
+            # - Store the artifact config in the step configuration of the
+            #   deployment. This makes sense in any case I think, also for the
+            #   comment below about linking models to the pipeline run.
+            # - Get rid of having a separate model for an output. Not sure what
+            #   the use case of that actually is and whether it warrants all
+            #   this complexity.
+            utils._link_cached_artifacts_to_model(
+                model_from_context=model,
+                step_run=step_run_request,
+                step_source=step_run.spec.source,
+            )
+            if model:
+                # TODO: When a step is not cached, we parse the output artifact
```

Comment:
@avishniakov

Reply:
For this and the other comment regarding output artifact configs, I think I don't really get the value of having a custom model version for a specific step output. Do we have any specific use-case in mind where this would be required?

If no, my gut feeling would be to just remove the option entirely, which would simplify things. Other than that, I guess we have to store the `ArtifactConfig` in the `StepConfiguration` and use it to link the artifacts to their models.

Reply:
Yes, let's just remove it. To the facts: I can imagine it is just a not implemeted use case, no good logic behind. I checked the code and cannot find better explanation.

Reply:
I've removed the option to specify the model version on the artifact config, and also stored the artifact config in the DB so it can be used to link the artifacts correctly as `model_artifact` or `deployment_artifact`